### PR TITLE
Introduce support for substring function - `str_sub()`

### PR DESCRIPTION
**Pull request by xSavitar**
This function takes in 3 arguments: the string, the position and the length (of the substring). As a quick example, a call to the function

`str_sub('c programming', 3, 11)`

will return the string `programming` as the substring to the string that is provisioned to the function call.

### CODE REVIEW

**Comment by xSavitar** [comment]
@ThakeeNathees, can you review this whenever you have time please. Thanks!

File: src/pk_core.c [hunk 1]
```diff
@@ -648,7 +648,38 @@ DEF(coreExit,
...
+  String* str = NULL;
+  if (!validateArgString(vm, 1, &str)) return;
+  if (!validateInteger(vm, ARG(2), &pos, "Argument 2")) return;
+  if (!validateInteger(vm, ARG(3), &len, "Argument 3")) return;
+
+  if (str->length < 1)
+    RET_ERR(newString(vm, "Expected a string length more than 0."));
+
+  if (str->length < pos || str->length < len)
+    RET_ERR(newString(vm, "Position or length out of range."));
```
Comment (ThakeeNathees) [inline src/pk_core.c]: Could you change the `Position or length ...` to `Index out of range.`

File: src/pk_core.c [hunk 2]
```diff
@@ -648,7 +648,38 @@ DEF(coreExit,
...
+  String* sub = NULL;
+  sub = ALLOCATE_DYNAMIC(vm, String, len + 1, char);
+
+  if (sub == NULL)
+    RET_ERR(newString(vm, "Unable to allocate memory."));
+
+  for (c = 0; c < len; c++) {
+    sub->data[c] = str->data[pos + c - 1];
+  }
+  sub->data[c] = '\0';
```
Comment (ThakeeNathees) [inline src/pk_core.c]: to create a new object, you have to use `new...` function inorder to initialize it and add the object to the VM's objects list. ex: to create a `list` use `newList()`

here you have to use the function  `String* sub = newStringLength(vm, str->data + pos, (uint32_t)len);`

File: src/pk_core.c [hunk 3]
```diff
@@ -648,7 +648,38 @@ DEF(coreExit,
...
+
+  RET(VAR_OBJ(newString(vm, sub->data)));
```
Comment (ThakeeNathees) [inline src/pk_core.c]: This can be returned as `RET(VAR_OBJ(sub))`

File: tests/lang/core.pk [hunk 4]
```diff
@@ -30,6 +30,9 @@ def fn(p1, p2, p3) end
 assert(fn.name == "fn")
 assert(fn.arity == 3)
 
+## String functions
+assert((str_sub('c programming', 3, 11)) == 'programming')
```
Comment (ThakeeNathees) [inline tests/lang/core.pk]: Our substring range should match like in other languages. (like in the example below)
```
substring('foobarbaz', 2, 3) = 'oba'

f o o b a r 
0 1 2 3 4 5
    ^---^
```

**Review by ThakeeNathees — CHANGES_REQUESTED** [review]

**Comment by xSavitar** [comment]
Thanks for the review @ThakeeNathees. I think latest patch should be good now. 💯

File: src/pk_core.c [hunk 5]
```diff
@@ -648,7 +648,37 @@ DEF(coreExit,
...
+  if (str->length < 1)
+    RET_ERR(newString(vm, "Expected a string length more than 0."));
+
+  if (str->length < pos || str->length < len)
+    RET_ERR(newString(vm, "Index out of range."));
+
+  String* subStr = newStringLength(vm, str->data + pos, (uint32_t)len);
+
+  if (subStr == NULL)
+    RET_ERR(newString(vm, "Unable to allocate memory."));
```
Comment (ThakeeNathees) [inline src/pk_core.c]: If there is an allocation failure, it should be handled at the VM's allocation function, otherwise we have to check if the allocation failed at every single place we're creating a new object. 

Remove these 2 lines.

File: src/pk_core.c [hunk 6]
```diff
@@ -648,7 +648,37 @@ DEF(coreExit,
...
+
+  for (c = 0; c < len; c++) {
+    subStr->data[c] = str->data[pos + c];
+  }
+  subStr->data[c] = '\0';
```
Comment (ThakeeNathees) [inline src/pk_core.c]: The `newStringLength()` will allocate a new string and copy each character using `memcpy()` from the given string. 

And we're passing `str->data + pos` this is where the substring would start, and the `newStringLength` will only copy the given `length` of characters.

Remove this.

File: src/pk_core.c [hunk 7]
```diff
@@ -648,7 +648,37 @@ DEF(coreExit,
 // String functions.
 // -----------------
 
-// TODO: substring.
+DEF(coreStrSub,
+  "str_sub(str:string, pos:num, len:num) -> string\n"
+  "Returns a substring from a given string supplied. In addition, "
+  "the position and length of the substring are provided when this "
+  "function is called. For example: `str_sub(str, pos, len)`.") {
+  int c = 0;
```
Comment (ThakeeNathees) [inline src/pk_core.c]: Since this variable `c` isn't going to be used, remove this and keep an empty line between the docstring and the function body.

File: src/pk_core.c [hunk 8]
```diff
@@ -648,7 +648,37 @@ DEF(coreExit,
...
+  int64_t len, pos;
+
+  String* str = NULL;
+  if (!validateArgString(vm, 1, &str)) return;
+  if (!validateInteger(vm, ARG(2), &pos, "Argument 2")) return;
+  if (!validateInteger(vm, ARG(3), &len, "Argument 3")) return;
+
+  if (str->length < 1)
+    RET_ERR(newString(vm, "Expected a string length more than 0."));
```
Comment (ThakeeNathees) [inline src/pk_core.c]: We should allow an empty string to be used to get substring too. otherwise, the user has to check if the string is empty or not every time they calling the substring function.

Also we allow the `length` to be `0` for the same reason (example below)

```c
  String* str = NULL;
  if (!validateArgString(vm, 1, &str)) return;
  if (!validateInteger(vm, ARG(2), &pos, "Argument 2")) return;
  if (!validateInteger(vm, ARG(3), &len, "Argument 3")) return;

  // validate pos and len values

  // End case.
  if (len == 0) RET(newStringLength(vm, "", 0));
```

File: tests/lang/core.pk [hunk 9]
```diff
@@ -30,6 +30,12 @@ def fn(p1, p2, p3) end
 assert(fn.name == "fn")
 assert(fn.arity == 3)
 
+## String functions
+assert(str_sub('c programming', 2, 11) == 'programming')
+assert(str_sub('foobarbaz', 2, 3) == 'oba')
+assert(str_sub('abcdef', 1, 2) == 'bc')
+assert(str_sub('I am a boy', 0, 4) == 'I am')
+
```
Comment (ThakeeNathees) [inline tests/lang/core.pk]: Looks good, but add ...
- empty string to the tests as well to cover all the edge cases.
- Also zero length substrings (`== ''`) 
- try the position equals to the string length and len = 0

**Review by ThakeeNathees — CHANGES_REQUESTED** [review]

File: src/pk_core.c (hunk 5 above)
Comment (ThakeeNathees) [reply src/pk_core.c]: Also, don't use `camelCase` for local variables, instead use `snake_case` like `sub_str` for the convention. See #150 for a minimal naming conventions pocketlang's using.

**Comment by xSavitar** [comment]
Thanks for the reviews. All request for changes applied 💯

File: src/pk_core.c [hunk 10]
```diff
@@ -648,11 +648,37 @@ DEF(coreExit,
...
+
+  if (str->length < pos || str->length < len)
+    RET_ERR(newString(vm, "Index out of range."));
```
Comment (ThakeeNathees) [inline src/pk_core.c:665]: Sorry I've missed it in the last review, We should allow the position range to be `0 <= pos <= str->length`
when the `pos == str->length` the user can still get a substring of length zero.

```c
if (pos < 0 || str->length < pos) RET_ERR(...);
```

The position should be added to the length check, if the length of the substring starting at the `pos` exceeded the string length it would be an error. (see below)

```c
if (str->length < pos + len) RET_ERR(...);
```

```
pos = 5
len = 7
                 v- str->length
"f o o b a r b a z"
           ^-----------^
           pos         pos+len
```


TODO:
I'm not expecting it from this pr but it's worth mentioning.
Make the length argument as optional to let the users allow get substring starting from `pos` to the reset of the string.

File: tests/lang/core.pk [hunk 11]
```diff
@@ -30,6 +30,15 @@ def fn(p1, p2, p3) end
...
 
+## String functions
+assert(str_sub('c programming', 2, 11) == 'programming')
+assert(str_sub('foobarbaz', 2, 3) == 'oba')
+assert(str_sub('abcdef', 1, 2) == 'bc')
+assert(str_sub('I am a boy', 0, 4) == 'I am')
+assert(str_sub('programming', 2, 0) == '')
+assert(str_sub('foobar', 5, 0) == '')
+assert(str_sub('foobar', 0, 6) == 'foobar')
+
```
Comment (ThakeeNathees) [inline tests/lang/core.pk:42]: add `assert(str_sub('', 0, 0) == '')` after removing the string length validation I've mentioned above.

**Review by ThakeeNathees — CHANGES_REQUESTED** [review]
Oops, I've forgotten to review regarding the range of the substring. Here it is.

File: src/pk_core.c (hunk 10 above)
Comment (xSavitar) [reply src/pk_core.c:665]: @ThakeeNathees, thanks for mentioning. I'll send it in a follow up patch please.

**Comment by ThakeeNathees** [comment]
@xSavitar Sorry, I've force pushed the master (which I shouldn't have done) Can you create a new PR for this or try to rebase this with the master?

**Comment by xSavitar** [comment]
Sure no problem!
